### ISSUE TITLE
cve-check-tool fails to run

### ISSUE THREAD

**Issue by ipuustin** [01org/meta-security-isafw#79]
After the recipe-specific-sysroots change was merged into oe-core, the cve-check-tool doesn't appear to run anymore. File `build/tmp-glibc/log/isafw-logs/isafw_cvelog` says:

    cve-check-tool is missing!

If I install cve-check-tool directly on the host, the error goes away. It could be that a different method for installing the tool is needed?

**Comment by ereshetova** [comment]
@ipuustin, could you please point to the change merged into oe-core? I wonder what happened...

**Comment by ipuustin** [comment]
@ereshetova the main patch is this: https://github.com/openembedded/openembedded-core/commit/809746f56df4b91af014bf6a3f28997d6698ac78

**Comment by ereshetova** [comment]
@ipuustin, this is a big change and I would need to talk with someone who understands the new setup well. 

So, the issue as far as I understood, is that previously cve-check-tool-native was installed once into a system root because it is needed by a task running the cve check. It is not any package dependency, but task dependency. In the new setup, there is no anymore any single sysroot, but many package specific ones. In our case, we still need cve-check-tool-native to be installed in every single of that sysroots in order for the whole thing to work (or maybe there is a new way to tell that this package is needed everywhere?). What probably happens is that it doesn't get installed into new separate sysroots.

Previously this is how dependency was declared:

https://github.com/01org/meta-security-isafw/blob/master/classes/isafw.bbclass#L26

Now, I don't know what is the correct way.

**Comment by mythi** [comment]
OE-core itself seems to suffer from the same problem http://lists.openembedded.org/pipermail/openembedded-core/2017-February/132268.html

**Comment by ereshetova** [comment]
@rpurdie, could you help to understand what is the new proper way to make sure a package is installed in all sysroots when it is needed for a certain task? 

Here, how dependency was declared before when it was still working and had a common root:

https://github.com/01org/meta-security-isafw/blob/master/classes/isafw.bbclass#L26

**Comment by rpurdie** [comment]
As far as I can tell, that dependency is correct. The main difference now would be that the tool is in the recipe specific native sysroot rather than in the global shared sysroot. Are you saying that cve-check-tool isn't anywhere in WORKDIR/recipe-sysroot* ?

**Comment by ereshetova** [comment]
this was my understanding: it cannot find the tool, but I don't have a setup. 

@jku, you have been debugging this locally: can you confirm that the tool is not present

**Comment by jku** [comment]
There are indeed several dependency issues at least in the oe-core cve-check-tool version: I'll catch up with you when I've figured those out (tomorrow hopefully).

**Comment by ereshetova** [comment]
ok, actually now when I finally reproduce the error locally, I figured the actual problem with the cve-check-tool in layer. I really forgot quite a lot of what was happening. 

So, cve-check-tool inside the layer is actually run only once for the whole set of image packages and it is done via process_reports_handler handler, which is attached to bb.event.BuildCompleted event.  

https://github.com/01org/meta-security-isafw/blob/master/classes/isafw.bbclass#L89

The problem that I don't understand in which context/sysroot this handler runs now after the sysroot change has been merged. Because what I need to tell in dependencies is to install the cve-check-tool into that particular sysroot. 

@rpurdie could you please suggest the correct dependency for it? And if this even possible in the new setup? I am having doubts on it...

**Comment by jku** [comment]
> So, cve-check-tool inside the layer is actually run only once for the whole set of image packages and it is done via process_reports_handler handler, which is attached to bb.event.BuildCompleted event.

Oh, I didn't realise that.

If I'm reading the class correctly do_analyse_image() uses the image sysroots: so `TMPDIR/work/<machine>/<image>/1.0-r0/recipe-sysroot-native/ `is where they should appear.

do_analyse_image has no dependencies, but there is this:
```
do_rootfs[depends] += "checksec-native:do_populate_sysroot ca-certificates-native:do_populate_sysroot"
do_rootfs[depends] += "prelink-native:do_populate_sysroot"
do_rootfs[depends] += "python-lxml-native:do_populate_sysroot"
```
I would have expected these to be do_analyse_image dependencies (although maybe I don't understand something here). More importantly there is no dependency on cve-check-tool-native:do_populate_sysroot -- I believe there should be.

**Comment by jku** [comment]
Also if the scanning is only done once per image, do_analysesource probably does not need the cve-check-tool and ca-certificates dependencies.

**Comment by ereshetova** [comment]
@jku, I tried putting cve-check-tool-native  dependency both in handler and in do_rootfs and with no end result change, so it is smth trickier here.

Also, handler is not run as part of do_analyse_image, but it is handler on build completed event, so I think this is why dependency to do_rootfs doesn't help.

**Comment by ereshetova** [comment]
But you are right that dependency to do_analysesource is not needed anymore, I will drop it.

**Comment by jku** [comment]
> handler is not run as part of do_analyse_image, but it is handler on build completed event

You are correct, sorry I misunderstood you at first. ImageSecurityAnalyser is used in so many places... I have no idea yet if there's a sysroot available at that point (doubt it). Is there a reason this cve-check-tool run can't happen at image build time?

**Comment by ereshetova** [comment]
>Is there a reason this cve-check-tool run can't happen at image build time?

Yes, because image is not always build, especially if you run things locally or in some cases you want a fast check. We had requirement in past that you are able to just check CVE status without image rebuild, so we have do_analysesource run per package very fast and then handler is triggered for cve checks at the end. This way we really can do scans super fast vs. when we rebuild the image, which is very slow.

**Comment by jku** [comment]
So there is a way to construct a sysroot when needed:

http://git.yoctoproject.org/cgit/cgit.cgi/poky-contrib/commit/?h=ross/index&id=86eea01b83628f5858be8c058bd4701b91491d6e

This would probably work for your handler as well.

**Comment by ereshetova** [comment]
Oh, looks nice, thanks! I am going to try.

**Comment by ereshetova** [comment]
Thanks to @jku, working now!

**Comment by ereshetova** [comment]
Closing this since issue is fixed
